Load text tracks from dash into video.js

justinanastos opened the pull request:
As of video.js v5.14, video.js no longer supports native subtitles on anything besides Safari. For Chrome, this means there is *no* subtitle support. This will wait for dash.js to load all the text tracks and then add them to video.js if we're using a version that does not attempt to use native subtitles. Audio tracks will also be added.


---

## Todo

- [x] Replace`.map` with `.forEach` https://github.com/videojs/videojs-contrib-dash/pull/135#pullrequestreview-18056224
- [x] Remove `WeakMap` https://github.com/videojs/videojs-contrib-dash/pull/135#pullrequestreview-18287005
- [x] Make sure to clean everything up when source is changed. See https://github.com/videojs/videojs-contrib-media-sources/pull/118#pullrequestreview-8644856
- [x] Don't use `Map`
- [x] Fix Safari
- [x] Update README.md
- [x] [Audio tracks looks good to me but I'm not seeing the text track cues populate for http://vm2.dashif.org/dash/vod/testpic_2s/multi_subs.mpd
](https://github.com/videojs/videojs-contrib-dash/pull/135#issuecomment-277347876)
- [x] ~~[Location for `automaticTextTrackSelection ` option](https://github.com/videojs/videojs-contrib-dash/pull/135#discussion_r99416456)~~

---

~~Wait for https://github.com/videojs/video.js/pull/4131 to be merged. Add that version as an explicit dependency (maybe change it to a `peerDependency` and `devDependency` since we don't actually need it here?~~

This fix depends on videojs@^v5.18

chemoish commented:
I might look into generating cues before calling `addRemoteTextTrack`, not sure if this will effect anything, but last time I was in `TextTrack` I did notice a few `setTimeout` and `Event`s that caused race conditions.

---

Somethings to consider in this implementation:

- How to manage the selected track? (haven't done a deep dive with managing initial track selection modes, but I did notice that embedded captions were being defaulted by the reference player)
- Apparently there are a few types of tracks that can be embedded, they can also have different roles, I am not sure how that would effect this implementation (since this looks like it only manages captions)

justinanastos commented:
@chemoish 

>I might look into generating cues before calling `addRemoteTextTrack`, not sure if this will effect anything, but last time I was in `TextTrack` I did notice a few `setTimeout` and `Event`s that caused race conditions (just to be aware).

I'm new to working with video.js's implementation, but the `addRemoteTextTrack` function appears to create the `HTMLTrackElement` element and I need that to add the `cue`s. 

Stepping through the `Html5` tech implementation related to `addRemoteTextTrack`, (https://github.com/videojs/video.js/blob/master/src/js/tech/html5.js#L712-L738), I didn't see any `Event`s or `setTimeout`s (at least not in v15.1). `addRemoteTextTrack` accepts an `options` object that is fed directly to the native [`HTMLTrackElement`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLTrackElement) which will optionally apply`kind`, `label`, `srclang`, `default`, and `src`; if any of those exist.

---

>How to manage the selected track? (haven't done a deep dive with managing initial track selection modes, but I did notice that embedded captions were being defaulted by the dash.js reference player)

I updated how I handle this. If I apply the `default` `option`, then both the defaulted track and "captions off" will be selected. I will loop through all the tracks after they have been generated and set the `mode` attribute to `"showing"` for `default: true` and `"hidden"` otherwise.

![](https://cl.ly/1D383R1H3A2j/Image%202017-01-18%20at%202.58.39%20PM.png)

---

>Apparently there are a few types of dash.js tracks that can be embedded, they can also have different roles, I am not sure how that would effect this implementation (since this looks like it only manages captions)—fyi, there isn't a 1:1 list of TextTracks from dash.js to HTML5 spec.

Agreed, the source for [`TextSourceBugger.js`](http://cdn.dashjs.org/latest/jsdoc/streaming_TextSourceBuffer.js.html#sunlight-1-line-151) appears to show that `dash.js` only supports captions and subtitles. Using the `kind` property that comes from the dash parsing and feeding that `kind` into `video.js` using this implementation will show two separate (and functional) captions and subtitles menus.

![](https://cl.ly/3e0H3z2U1g1Z/Image%202017-01-18%20at%203.09.01%20PM.png)

justinanastos commented:
I've reviewed all the samples in #103. This PR in it's current state does not match the functionality of the dash reference player. I'll keep working on it.

justinanastos commented:
With this update, it's worked with everything I can throw at it @chemoish .

justinanastos commented:
I found what I perceive to be a flaw in the text track selection logic. This assumes a 1:1 relationship between dash text tracks and videojs's DOM track list. This is not guaranteed because the user can manipulate the videojs text tracks however they like (or not even use them), so this needs to be modified.

chemoish commented on src/js/setup-text-tracks.js:
```diff
@@ -0,0 +1,125 @@
+      // Example:
+      // https://storage.googleapis.com/shaka-demo-assets/sintel-mp4-wvtt/dash.mpd
+
+      // Return `track` so we can continue chaning.
```
`chaining` + does this require chaining still?

justinanastos replied:
@chemoish No, it doesn't :)

chemoish replied:
possibly update to `.each` and remove comment/return.

chemoish commented on src/js/videojs-dash.js:
```diff
@@ -89,14 +91,25 @@ class Html5DashJS {
+
+    // When `dashjs` finishes loading metadata, create audio tracks for
+    // `video.js`.
+    this.mediaPlayer_.on(
```
minor, any reason why the event handler is bound inline rather than within the setup?

justinanastos replied:
No good reason @chemoish , I will update it to make them consistent.

justinanastos commented:
Updates pushed with the following changes:

- Text track handling is no longer async because I am passing `tech` into `setupTextTracks` and it can then check for `tech.featuresNativeTextTracks`.
- `setupTextTracks` is called in the same fashion as `setupAudioTracks`
- Text track handling is no longer dependent on the dash text track list being a 1:1 match for the videojs track list because users can destroy or create text tracks at will. This uses a `WeakMap` as a dictionary.

gkatsev commented on src/js/setup-text-tracks.js:
```diff
@@ -0,0 +1,136 @@
+function attachDashTextTracksToVideojs(player, tracks) {
+  const mediaPlayer = player.dash.mediaPlayer;
+
+  const trackDictionaryKeyedByTextTrack = new WeakMap();
```
I don't think it's a good idea to use WeakMap yet, still. To support the widest field of browsers possible, we can't use WeakMap and WeakMap isn't really polyfillable.

justinanastos replied:
I'll replace it with a `Map()`

gkatsev replied:
We might not want to either Map either depending on which browsers we're targeting for this project. We wouldn't want to require es6-shim for `Map()` support if we expect to be able to run this package in a browser that supports MSE+EME but not `Map()`.

@forbesjo do need to worry about this?

justinanastos replied:
My mistake, I didn't check to see how the babel transforms were handled. If you're not polyfilling in `Map`, then I agree it's unacceptable. I'll replace it.

gesinger commented on src/js/setup-text-tracks.js:
```diff
@@ -0,0 +1,136 @@
+  */
+  const textTracks = player.textTracks();
+
+  for (let i = 0; i < textTracks.length; i += 1) {
```
`i++` just for consistent style

gesinger commented on src/js/setup-text-tracks.js:
```diff
@@ -0,0 +1,136 @@
+  mediaPlayer.updateDashTextTrack();
+
+  // Update dash when videojs's selected text track changes. Use an arrow function so
+  // `mediaPlayer.updateDashTextTrack` can be overriden exterally.
```
I'm not sure if we are ready yet to provide an external function that people can override. If we do though, we may want to add it onto player.dash, since mediaPlayer is technically a dash.js object that we don't have control over.

justinanastos replied:
This was a mistake, I had used this logic elsewhere and forgot to remove it. I'll remove it here.

justinanastos commented on src/js/setup-text-tracks.js:
```diff
@@ -0,0 +1,136 @@
+  tracks
+    // Map input data to match HTMLTrackElement spec
+    // https://developer.mozilla.org/en-US/docs/Web/API/HTMLTrackElement
+    .map((track) => Object.assign(
```
No good reason, I can axe the `track` reference.

justinanastos commented on src/js/setup-text-tracks.js:
```diff
@@ -0,0 +1,136 @@
+    const textTracks = this.textTracks();
+    let activeTextTrackIndex = -1;
+
+    function findMatchingTrack(dashTrack) {
```
That was my original plan, but jshint didn't like me creating functions within a loop.

justinanastos commented on src/js/setup-text-tracks.js:
```diff
@@ -0,0 +1,136 @@
+    // Add track to videojs track list
+    .map((track) => {
+      const remoteTextTrack = player.addRemoteTextTrack(track, true);
+      trackDictionaryKeyedByTextTrack.set(remoteTextTrack.track, track);
```
I'm not sure I understand. Later on I use index matching because I have to loop through `player.textTracks()` to find the active track. After that, I use that track to figure out the dash track. `dash.setTextTrack()` expects an index so I have to first find it, then figure out it's index. I must be missing something?

justinanastos replied:
Great point. When we're using remote time text files `.getTracksFor('text')` will return an empty array so we have no choice but to use the `tracks` passed in the event handler.

justinanastos commented:
Updated @gesinger , @gkatsev , and @chemoish .

justinanastos commented:
Use of `Map` has been removed.

justinanastos commented:
The audio track manipulation breaks dash playback in the 10.0.3. I am investigating.

forbesjo commented on src/js/videojs-dash.js:
```diff
@@ -89,6 +91,12 @@ class Html5DashJS {
     this.mediaPlayer_.attachSource(manifestSource);
 
+    // Setup text tracks
+    setupTextTracks.call(null, this.player, tech);
```
Should these be called before attachSource()?

forbesjo replied:
My concern is that `dashjs.MediaPlayer.events.TEXT_TRACKS_ADDED`/`dashjs.MediaPlayer.events.PLAYBACK_METADATA_LOADED` may fire between `attachSource()` and the event listener setup functions

justinanastos replied:
This makes sense. If the user can fetch a source and attach it, this might become a sync call and the events might get missed. I've moved the calls to be before `attachSource()`.

justinanastos commented:
I have a question and could use some advice. I'm having trouble bypassing native captions in Safari because of this line:

https://github.com/videojs/video.js/blob/master/src/js/tech/html5.js#L798

I have been unsuccessful in finding a smooth way to disable native text tracks other than globally disabling it in the Html5 tech when `videojs-contrib-dash` is included, but this might break other techs/plugins. The only solution I have found that seems to work is to force the user to pass in `{html: {nativeCaptions: false}}` in their `options` when instantiating `videojs`.

There has to be a better way, can anyone give me some advice on where to look?

forbesjo commented:
We've encountered a similar scenario when working with videojs-contrib-hls and Edge's native captions and have come to the same conclusion that the user needs to disable `nativeCaptions`. @gkatsev is this something that can be solved with middleware?

forbesjo commented:
@justinanastos any objection to merging and releasing this for now and taking care of Safari later?

gesinger commented on src/js/setup-text-tracks.js:
```diff
@@ -0,0 +1,144 @@
+      const remoteTextTrack = player.addRemoteTextTrack(trackConfig, true);
+      trackDictionary.push({textTrack: remoteTextTrack.track, dashTrack});
+
+      // Don't add the cues becuase we're going to let dash handle it natively. This will ensure
```
Minor, because

gesinger commented on src/js/setup-text-tracks.js:
```diff
@@ -0,0 +1,144 @@
+ * @private
+ */
+function attachDashTextTracksToVideojs(player, tech, tracks) {
+  const trackDictionary = [];
```
Minor, but may want to change the name since it is an array. Maybe something like trackPairs?

justinanastos commented:
@forbesjo I feel like I'm very close to having Safari working flawlessly as well. If I don't have it by tomorrow, I'll take it out of the scope.

justinanastos commented:
@forbesjo I have Safari working now. I'll clean the code up and update the PR tomorrow.

justinanastos commented:
This now works with Safari and I've updated the readme. This is ready for another review.

forbesjo commented:
Audio tracks looks good to me but I'm not seeing the text track cues populate for http://vm2.dashif.org/dash/vod/testpic_2s/multi_subs.mpd

forbesjo commented on README.md:
```diff
@@ -77,6 +77,34 @@ var updateSourceData = function(source) {
+videojs('example-video', {
+  html5: {
+    nativeCaptions: false,
+  },
```
Trailing commas

justinanastos replied:
Fixed.

forbesjo commented on README.md:
````diff
@@ -77,6 +77,34 @@ var updateSourceData = function(source) {
+```javascript
+videojs('example-video', {
+  html5: {
+    automaticTextTrackSelection: false,
````
Looks like this lives in `html5.dash`

justinanastos replied:
I opted to _not_ add it to `html5.dash` because that's where the options for `dashjs` are stored. This is what I was referencing in https://github.com/videojs/videojs-contrib-dash/pull/144#issuecomment-277023964 when I suggested it might be beneficial to have the options passed to `dashjs` in a separate object like `html5.dash.dashjsOptions`.

https://github.com/videojs/videojs-contrib-dash/pull/135/files#diff-a10f3c20a3b12e434e950fa44ed404d2R49

https://github.com/videojs/videojs-contrib-dash/pull/135/files#diff-a10f3c20a3b12e434e950fa44ed404d2R49

forbesjo replied:
Oh I was looking at your filter at https://github.com/videojs/videojs-contrib-dash/pull/135/files#diff-5b161687541b3fff374a3771055376aaR86

I'm not understanding the usecase for this option could you walk me through the scenario?

justinanastos replied:
Good point, I forgot I had the filter in there.

My use case is I have video that has 4 text tracks, 2 for English and 2 for an alternate language. One English and one alternate language track for forced narratives. I'm having this PR add all 4 tracks to videojs, then I go in after and remove the forced narratives. When captions are turned off, I determine the current audio language and have dash display the corresponding forced narrative track. That's why I want `videojs-contrib-dash` to load the tracks but not handle switching them.

forbesjo replied:
Ok so you only want to show the user 2 of the 4 text tracks and handle the other two yourself?

forbesjo replied:
If this is the case you could do something like this:

```js
var menu = player.controlBar.captionsButton.menu;

menu.children()
  .filter(captionsToHide);
  .forEach(function(item) {
    menu.removeChild(item);
  });

var offMenuItem = menu.children().filter(findOffMenuItem)[0];

offMenuItem.on('change', detectAudioTrackAndTurnOnCaption);
```

justinanastos replied:
That is very close to what I'm doing, my logic was to prevent both `videojs-contrib-dash` and my application from trying to set the active dash track. I don't know if that's necessary though.

justinanastos commented:
@forbesjo 
>Audio tracks looks good to me but I'm not seeing the text track cues populate for http://vm2.dashif.org/dash/vod/testpic_2s/multi_subs.mpd
When you say "text track cues populate", do you mean they don't show up or that video.js's addCues is never called?

forbesjo commented:
I wonder if vtt.js is conflicting with how dash.js is handling the tracks, it works for me when I use the video.novtt.js script.

forbesjo commented:
Perhaps the audio track work should be split into a separate PR since that seems to be working fine. 

The README and code should also be updated to mention using `html5.nativeAudioTracks=false` to support browsers with native audio track APIs like Edge.

forbesjo commented:
@justinanastos I'm going to open a PR for audio tracks (251c486) and add some tests.

justinanastos commented:
@forbesjo I'm investing how to use `html5.nativeAudioTracks` and I'd be happy to split the PR into separate audio/text PRs.

justinanastos commented:
@forbesjo I am failing to understand what you said about using `html.nativeAudioTracks`, can you give me any more information?

forbesjo commented:
If the browser supports native audio tracks video.js will automatically hook into `videoEl.audioTracks`, setting `html.nativeAudioTracks` to `false` will have a similar effect to `html.nativeTextTracks` where it will ignore whatever the native API is doing. 

Dash.js manages the audio tracks internally so on browsers like Edge/Safari using video.js `player.audioTracks()` will only return one track instead of the appropriate amount.

justinanastos commented:
@forbesjo 

>I wonder if vtt.js is conflicting with how dash.js is handling the tracks, it works for me when I use the video.novtt.js script.

Still trying to get the bottom of this.

justinanastos commented:
@forbesjo If [videojs-vtt.js](https://www.npmjs.com/package/videojs-vtt.js) isn't included, this apparently doesn't work in Chrome.

[This line](https://github.com/Dash-Industry-Forum/dash.js/blob/development/src/streaming/TextTracks.js#L480) will throw :

```
"Failed to execute 'addCue' on 'TextTrack': parameter 1 is not of type 'TextTrackCue'."

"TypeError: Failed to execute 'addCue' on 'TextTrack': parameter 1 is not of type 'TextTrackCue'.
    at Object.addCaptions (http://localhost:9090/player.js:49535:20)
    at Object.addTextTrack (http://localhost:9090/player.js:49222:23)
    at createTextTrackFromMediaInfo (http://localhost:9090/player.js:46662:25)
    at Object.append (http://localhost:9090/player.js:46752:18)
    at http://localhost:9090/player.js:46330:42
    at waitForUpdateEnd (http://localhost:9090/player.js:46400:14)
    at Object.append (http://localhost:9090/player.js:46327:14)
    at onInitFragmentLoaded (http://localhost:9090/player.js:60045:33)
    at http://localhost:9090/player.js:44623:38
    at Array.forEach (native)"
__proto__
```

Am I fair to assume that this is because there is an in complete shim in video.js? Therefore, we would need to require the user to use `video.novtt.js` or to include `videojs-vtt.js`?

justinanastos commented:
@forbesjo I updated the readme with the following:

>You must either use `video.novtt.js` or include [`videojs-vtt.js`](https://www.npmjs.com/package/videojs-vtt.js) to have `videojs-contrib-dash` support captions.

If the user doesn't do either of those, captions won't work but playback will not be effected.

forbesjo commented:
The video.js script should include videojs-vtt.js automatically, a user shouldn't have to think about videojs-vtt.js

justinanastos commented:
@forbesjo  I'm at a total loss as to why I'm seeing this error. https://github.com/videojs/videojs-contrib-dash/pull/135#issuecomment-278427696

forbesjo commented:
I'll dig into it tomorrow

justinanastos commented:
@forbesjo and @chemoish I want to give a summary of what I'm seeing because I don't know how to fix it.

# Situation: On Chrome, `video.js` includes `videojs-vtt.js` on it's own

- If `window.WebVTT` doesn't exist, `video.js` will include it and apply all the shims on it to `window`.

    ```
    if (!window.WebVTT && this.el().parentNode !== null && this.el().parentNode !== undefined) {
      const vtt = require('videojs-vtt.js');

      // load via require if available and vtt.js script location was not passed in
      // as an option. novtt builds will turn the above require call into an empty object
      // which will cause this if check to always fail.
      if (!this.options_['vtt.js'] && isPlain(vtt) && Object.keys(vtt).length > 0) {
        Object.keys(vtt).forEach(function(k) {
          window[k] = vtt[k];
        });
        this.trigger('vttjsloaded');
        return;
      } 

      ...
    ```

    [video.js#src/js/tech/tech.js#L532-544](https://github.com/videojs/video.js/blob/v5.17.0/src/js/tech/tech.js#L532-L544)

- Dash will instantiate cuepoints with `window.VTTCue`

    ```
    Cue = window.VTTCue || window.TextTrackCue;
    ```

    [dashjs#src/streaming/TextTracks.jsL61](https://github.com/Dash-Industry-Forum/dash.js/blob/development/src/streaming/TextTracks.js#L61)

    ```
    cue = new Cue(currentItem.start - timeOffset, currentItem.end - timeOffset, currentItem.data);
    ```

    [dashjs#src/streaming/TextTracks.jsL463](https://github.com/Dash-Industry-Forum/dash.js/blob/development/src/streaming/TextTracks.js#L463)

- Dash will throw:

    ```
    Failed to execute 'addCue' on 'TextTrack': parameter 1 is not of type 'TextTrackCue'.
    ```

- After throwing, playback will function as expected but no text tracks will be added to dash and, therefore, the `TEXT_TRACKS_ADDED` event will not be emitted so `videojs-contrib-dash` won't get any tracks either.

## Summary

It appears as though using a combination of a shim and native code on Chrome is causing an issue. If I prevent the shimming of `video-vtt.js`, then the native tracks work as expected.

---

I would like some help on how to proceed.

justinanastos commented:
@chemoish and @forbesjo : Maybe this is a bug in `dashjs`? Maybe the text tracks should default to a native implementation instead of WebVTT?

justinanastos commented:
I found that the bug mentioned before *does* break playback with fragmented text captions.

justinanastos commented:
@forbesjo I found a solution to the conflict with dashjs and videojs. It's not elegant, but it works exactly as we'd expect. The problem I ran into with Chrome is that video.js was overriding `window.VTTCue` with it's own version. Chrome will throw if you attempt to call `textTrack.addCue()` with anything besides it's own native `VTTCue`. I submitted a PR to video.js (videojs/video.js#4094) but it was declined for fear it would break break emulated text tracks.

[This fix](https://github.com/videojs/videojs-contrib-dash/pull/135/commits/0c24dc5189bdd23497c9933b44b090dde3366ba9) will prevent videojs from applying it's own version of `vtt.js` to `window`, allowing Chrome to function as expected. This will eliminate emulated text tracks on all browsers that support Dash, but this shouldn't be an issue because I believe all those browsers support native text tracks as well.

justinanastos commented:
@forbesjo I have removed the `automaticTextTrackSelection` option as well. I thought it was needed before I discovered https://github.com/Dash-Industry-Forum/dash.js/issues/1773. The fix has been merged (https://github.com/Dash-Industry-Forum/dash.js/pull/1774). When that is released, users can then override the default behavior.

justinanastos commented:
Sorry to do this again, but this appears to not work correctly if text tracks are removed from video.js, I'm working on it.

justinanastos commented:
Updated this PR to work when tracks are removed. The final commit can be removed when https://github.com/videojs/video.js/pull/4115 is merged and a new PR is merged against `video.js@5.x`.

gkatsev commented:
https://github.com/videojs/video.js/pull/4131 has been merged and released as videojs 5.18.0 and `next` on npm.

justinanastos commented:
I've removed the `vtt.js` hack and updated `dependencies` to require `video.js@^v5.18.0`

justinanastos commented:
Anything else I can do here @forbesjo ?

forbesjo commented:
I'll test this out and merge it in once 5.18.0 is `latest`

jinpark commented:
looks like 5.18 is latest now :)

forbesjo commented:
Nice, I'll see if I can get this merged in and cut a version this week.

forbesjo commented:
I haven't forgotten about this PR, just other tasks have put this on the back burner at the moment. This will get tested/merged sometime soon.

justinanastos commented:
I hate to say this, but I found a bug. `tech.clearTracks(['text']);` is unacceptable because it throws away all tracks, even ones added by something besides dash.

justinanastos commented:
I've updated the MR to stop using `tech.clearTracks(['text']);` and instead keep a list of tracks we've added and only remove those during cleanup. Save to merge again.

justinanastos commented:
I've been making updates and force pushing. I'm reverting them and going to add new commits that I will tag with `fixup: ` and then rebase when this is approved.

justinanastos commented:
Any updates on this @forbesjo ?

squarebracket commented:
FYI, I downloaded the patch of this PR, and noticed that changing the caption display settings doesn't actually change them.

justinanastos commented:
@squarebracket 

>FYI, I downloaded the patch of this PR, and noticed that changing the caption display settings doesn't actually change them.

That's correct. The captions will be displayed natively by dash.js.

forbesjo commented:
@justinanastos testing this now

squarebracket commented:
@justinanastos Why would that preclude video.js from setting the caption style? Aren't the dash.js captions being spit out as `VTTCue`s?

gkatsev commented:
Dash.js renders the captions natively rather than using Video.js's rendering capabilities. That means that the styling that Video.js can provide via the caption settings menu isn't (and potentially can't be) applied.
We probably should disable that menu item in this case.

squarebracket commented:
Can't videojs just set CSS classes or something though?

forbesjo commented:
@justinanastos your PR works for me but as has been discussed the captions style cannot be changed since dash.js is doing the rendering and not video.js. Looking into this

forbesjo commented:
It doesn't look like there's a mechanism to disable the dash.js display so I agree with @gkatsev that the caption settings menu item should be disabled.

squarebracket commented:
If they're going to be left completely to dash.js, then may I suggest doing something this in `Html5DashJS`:

```javascript
    // Attach the source with any protection data
    this.mediaPlayer_.setProtectionData(this.keySystemOptions_);
    this.mediaPlayer_.attachSource(manifestSource);
    // Use the normal text track display to render TTMLs into
    this.mediaPlayer_.attachTTMLRenderingDiv($('.vjs-text-track-display')[0]);

    this.tech_.triggerReady();
```

Their TTML rendering is much better than their VTTCue rendering. For instance, it can do this:
![screenshot from 2017-05-02 21-52-30](https://cloud.githubusercontent.com/assets/127774/25857618/a2396e28-34a7-11e7-9c09-4bbadf65aeb2.png)
And also supports vertical placement which is important for things like news shows:
![screenshot from 2017-05-09 10-38-49](https://cloud.githubusercontent.com/assets/127774/25857652/baae4352-34a7-11e7-8474-ad5710e71f44.png)

squarebracket commented:
In case anyone was testing media with CEA608 captions, I just had some fixes merged into dash.js. They should now work properly. (For reference, [here](https://github.com/Dash-Industry-Forum/dash.js/pull/1918) [here](https://github.com/Dash-Industry-Forum/dash.js/pull/1905) and [here](https://github.com/Dash-Industry-Forum/dash.js/pull/1907).)

forbesjo commented:
@squarebracket I'm not sure if it's a great idea to reuse video.js's text track display for TTML rendering. Instead this script can add a new component div to the player specifically for TTML. That can be done in a separate PR.

@gkatsev what's the recommended way of disabling menu items like the captions settings item? Thinking about this a little more we would still want to display caption settings if there are side-loaded captions (captions not coming from the source manifest). Though when a mix of in-manifest and external captions are present in the player a view might be confused when some captions can be customized while others cannot.

forbesjo commented:
I think we should leave the captions settings alone, merge this PR in and add other enhancements later

justinanastos commented:
@squarebracket I believe native cue rendering can do the same thing you're showing, right? There can be multiple cues on the screen at once and their position is defined by the vtt file.

squarebracket commented:
@justinanastos that is correct, however that information is dropped in the dash.js' VTTCue rendering library, but is rendered for the TTML rendering library.
